### PR TITLE
Implements fdk_aac decoder

### PR DESCRIPTION
**Pull request by liushuyu**
This PR adds an alternative AAC decoder which uses `fdk_aac` library directly.

This is for those whose `ffmpeg` is broken or don't want to install `ffmpeg` or Windows is too old to support Media Foundation.

Also, probably fixes the issues Linux Flatpak is currently experiencing with the AAC decoding.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4764)

### CODE REVIEW

**Comment by liushuyu** [comment]
CI issues are caused by lacking dependency (`fdk_aac`)

File: src/audio_core/hle/fdk_decoder.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2019 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <fdk-aac/aacdecoder_lib.h>
```
Comment (lioncash) [inline src/audio_core/hle/fdk_decoder.h]: This should likely go in the cpp file so the fdk includes aren't dumped into whatever includes this header (nothing in the header uses anything from it anyways)

Comment (liushuyu) [reply src/audio_core/hle/fdk_decoder.h]: Fixed in dd9ecd98b6f971def0c54c8b3140e6996624869a

Comment (liushuyu) [reply src/audio_core/hle/fdk_decoder.h]: Marking as resolved

**Comment by liushuyu** [comment]
Bad `fdk_aac` detection implemented. 

However, it's decided by whether `SBC` decoder is present because you want to remove it if you want to create a patent-free version, you need to get rid of all the patented coefficients stored inside that decoder.

**Comment by zhaowenlan1779** [comment]
What's the status of this?

**Comment by liushuyu** [comment]
>What's the status of this?

Sorry for the late response, the PR is ready despite the CI errors. If the CI errors were to be fixed, `ffmpeg` decoder wouldn't be tested on the Linux system at the current state (You can always add new containers to test different configurations, however).

For those who don't fully understand why this PR exists:

>Also, probably fixes the issues Linux Flatpak is currently experiencing with the AAC decoding.

Flatpak does not ship a working `ffmpeg` or `fdk_aac` (with AAC decoder stripped out due to patent issues of course) but VLC ships a working `fdk_aac` with one of its plugins which could be imported into other Flatpaks using some configuration tricks.

**Comment by B3n30** [comment]
ci failures

File: src/audio_core/hle/fdk_decoder.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+    std::optional<BinaryResponse> Decode(const BinaryRequest& request);
+
+    void Clear();
+
+    int GetFDKInfo();
+
+    Memory::MemorySystem& memory;
+
+    HANDLE_AACDECODER decoder = NULL;
```
Comment (zhaowenlan1779) [inline src/audio_core/hle/fdk_decoder.cpp]: Can `= nullptr` or `{}` be used here?

File: src/audio_core/hle/fdk_decoder.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    // choose the input format when initializing: 1 layer of ADTS
+    decoder = aacDecoder_Open(TRANSPORT_TYPE::TT_MP4_ADTS, 1);
+    // set maximum output channel to two (stereo)
+    // if the input samples have more channels, fdk_aac will perform a downmix
+    AAC_DECODER_ERROR ret = aacDecoder_SetParam(decoder, AAC_PCM_MAX_OUTPUT_CHANNELS, 2);
+    if (ret != AAC_DEC_OK) {
+        // unable to set this parameter reflects the decoder implementation might be broken
+        // we'd better shuts down everything
+        aacDecoder_Close(decoder);
+        decoder = NULL;
```
Comment (zhaowenlan1779) [inline src/audio_core/hle/fdk_decoder.cpp]: Ditto regarding the `NULL`

**Comment by zhaowenlan1779** [comment]
@liushuyu 
> Flatpak does not ship a working `ffmpeg` or `fdk_aac` (with AAC decoder stripped out due to patent issues of course) but VLC ships a working `fdk_aac` with one of its plugins which could be imported into other Flatpaks using some configuration tricks.

Then is it possible/feasible to implement such configuration tricks in Citra so that Flatpak users can benefit from HLE AAC audio?

File: src/audio_core/hle/fdk_decoder.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
```
Comment (zhaowenlan1779) [inline src/audio_core/hle/fdk_decoder.h]: This is back here again somehow

Comment (liushuyu) [reply src/audio_core/hle/fdk_decoder.h]: >>We do not separate includes

>This is back here again somehow

Well, I think you suggested me to move this include to here 6 months ago.

File: src/audio_core/hle/fdk_decoder.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            // get the stream information
+            stream_info = aacDecoder_GetStreamInfo(decoder);
+            // fill the stream information for binary response
+            response.num_channels = stream_info->aacNumChannels;
+            response.num_samples = stream_info->frameSize;
+            // fill the output
+            // the sample size = frame_size * channel_counts
+            for (int sample = 0; sample < (stream_info->frameSize * 2); sample++) {
+                for (int ch = 0; ch < stream_info->aacNumChannels; ch++) {
+                    out_streams[ch].push_back(decoder_output[(sample * 2) + 1]);
```
[the remark this answers is not in the document]
Comment (liushuyu) [reply src/audio_core/hle/fdk_decoder.cpp]: I observed other decoders use a similar pattern here: https://github.com/citra-emu/citra/blob/59c159e8a072d8bfa93161bd2219482f29a2a3f4/src/audio_core/hle/ffmpeg_decoder.cpp#L221

File: src/audio_core/hle/fdk_decoder.h (hunk 4 above)
Comment (zhaowenlan1779) [reply src/audio_core/hle/fdk_decoder.h]: Well, maybe you didn't understand me correctly, this include should definitely be in the cpp file as we do not want to expose fdk headers. By saying 'We do not separate includes' I mean there shouldn't be empty lines between includes.

Comment (liushuyu) [reply src/audio_core/hle/fdk_decoder.h]: >Well, maybe you didn't understand me correctly, this include should definitely be in the cpp file as we do not want to expose fdk headers. By saying 'We do not separate includes' I mean there shouldn't be empty lines between includes.

Oh, sorry about that, I will fix that. Is there any other issue I need to address?

**Pull request merged**